### PR TITLE
[odoo] [MIG] membership: migrated into new api

### PR DESCRIPTION
**Pull request by mag-odoo**
<strong>Task:</strong>https://www.odoo.com/web#id=13366&view_type=form&model=project.task&action=327
<strong>Pad:</strong>https://pad.odoo.com/p/r.S5uGyIoj6iUUk8R0

### CODE REVIEW

File: addons/membership/models/invoice.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    _inherit = 'account.invoice'
+   
+    @api.multi
+    def action_cancel(self):
+        '''Create a 'date_cancel' on the membership_line object'''
+        MembershipLine = self.env['membership.membership_line']
+        today_date = date.today()
+        for invoice in self:
+            mlines = MembershipLine.search([('account_invoice_line_id', 'in',
+                                            [l.id for l in invoice.invoice_line_ids])])
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: instead of doing every time search for membershipline, you can directly do like this way: 

invoice_line_ids = self.mapped('invoice_line_ids').ids
MembershipLine.search([('account_invoice_line_id', 'in', invoice_line_ids)]). no need of for loop.

File: addons/membership/models/invoice.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                    MembershipLine.unlink(ml_ids)
+        return res
+    
+    @api.multi
+    def unlink(self):
+        """Remove Membership Line Record for Account Invoice Line
+        """
+        MembershipLine = self.env['membership.membership_line']
+        for membership_line in MembershipLine.search([('account_invoice_line_id', 'in', self.ids)]):
+            membership_line.unlink()
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: why iteration ? you can directly unlink with recordsets of membership lines
membership_lines = MembershipLine.search([('account_invoice_line_id', 'in', self.ids)])
membership_lines.unlink()

File: addons/membership/models/invoice.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class Invoice(models.Model):
+    _inherit = 'account.invoice'
+   
+    @api.multi
+    def action_cancel(self):
+        '''Create a 'date_cancel' on the membership_line object'''
+        res = super(Invoice, self).action_cancel()
+        today_date = date.today()
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: Use fields.Date.today() ;)

File: addons/membership/models/invoice.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+   
+    @api.multi
+    def write(self, vals):
+        res = super(AccountInvoiceLine, self).write(vals)
+        MembershipLine = self.env['membership.membership_line']
+        for line in self.filtered(lambda l: l.invoice_id.type == 'out_invoice'):
+            membership_lines = MembershipLine.search([('account_invoice_line_id', '=', line.id)])
+            if line.product_id and line.product_id.membership and not membership_lines:
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: No need to check this "line.product_id and"  in if condition simply : 
if line.product_id.membership and not membership_lines:

File: addons/membership/models/invoice.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                # Product line has changed to a membership product
+                date_from = line.product_id.membership_date_from
+                date_to = line.product_id.membership_date_to
+                if line.invoice_id.date_invoice > date_from and line.invoice_id.date_invoice < date_to:
+                    date_from = line.invoice_id.date_invoice
+                MembershipLine.create({
+                                'partner_id': line.invoice_id.partner_id.id,
+                                'membership_id': line.product_id.id,
+                                'member_price': line.price_unit,
+                                'date': date.today(),
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: fields.Date.today()

File: addons/membership/models/invoice.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        membership_lines = self.env['membership.membership_line'].search([('account_invoice_line_id', 'in', self.ids)])
+        membership_lines.unlink()
+        return super(AccountInvoiceLine, self).unlink()
+
+    @api.model
+    def create(self, vals):
+        MembershipLine = self.env['membership.membership_line']
+        result = super(AccountInvoiceLine, self).create(vals)
+        membership_lines = MembershipLine.search([('account_invoice_line_id', '=', result.id)])
+        if result.invoice_id.type == 'out_invoice' and result.product_id and result.product_id.membership and not membership_lines:
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: when we got recordset then no need to check ifcondition for recordset. 

remove "result.product_id and" from above if condition

File: addons/membership/models/invoice.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            # Product line is a membership product
+            date_from = result.product_id.membership_date_from
+            date_to = result.product_id.membership_date_to
+            if result.invoice_id.date_invoice > date_from and result.invoice_id.date_invoice < date_to:
+                date_from = result.invoice_id.date_invoice
+            MembershipLine.create({
+                        'partner_id': result.invoice_id.partner_id and result.invoice_id.partner_id.id or False,
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: simply :  'partner_id': result.invoice_id.partner_id.id ;)

File: addons/membership/models/membership.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    _name = 'membership.membership_line'
+    _description = __doc__
+    _rec_name = 'partner_id'
+    _order = 'id desc'
+    partner_id = fields.Many2one('res.partner', 'Partner', ondelete='cascade', index=1, oldname='partner')
+    membership_id = fields.Many2one('product.product', string="Membership", required=True)
+    date_from = fields.Date('From', readonly=True)
+    date_to = fields.Date('To', readonly=True)
+    date_cancel = fields.Date('Cancel date')
+    date = fields.Date('Join Date', help="Date on which member has joined the membership")
```
Comment (atp-odoo) [inline addons/membership/models/membership.py]: change field name to "date_join"  with keep oldname="date" and make separate commit too....

File: addons/membership/models/membership.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+            state = 'none'
+            if (istate == 'draft') | (istate == 'proforma'):
+                state = 'waiting'
+            elif istate == 'open':
+                state = 'invoiced'
+            elif istate == 'paid':
+                state = 'paid'
+                for payment in line.account_invoice_id.payment_ids:
+                    if payment.invoice_ids and any(inv.type == 'out_refund' for inv in payment.invoice_ids):
+                        state = 'canceled'
```
Comment (atp-odoo) [inline addons/membership/models/membership.py]: this three lines: 56 to 58 can be done in single line using mapped and filtered function like this way : 
if line.account_invoice_id.mapped(payment_ids.invoice_ids).filtered(lambda i: i.type == 'out_refund'):
         state = 'canceled'
no need of iteration...

File: addons/membership/models/membership.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            elif istate == 'cancel':
+                state = 'canceled'
+            line.state = state
+
+    @api.multi
```
Comment (atp-odoo) [inline addons/membership/models/membership.py]: no need to declare @api.multi. constrains decorator is enough.

File: addons/membership/models/membership.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                ml.account_invoice_line_id = ail.id
+                )
+            JOIN account_invoice ai ON (
+            ai.id = ail.invoice_id)
+            WHERE ml.id IN %s''', (tuple(self.ids),))
+        records = self.env.cr.fetchall()
+        for record in records:
+            if record[0] and record[0] < 0:
+                raise UserError(_('Error, this membership product is out of date'))
+        return
```
Comment (atp-odoo) [inline addons/membership/models/membership.py]: no need of return......

File: addons/membership/models/partner.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+from odoo.exceptions import UserError
+
+
+class Partner(models.Model):
+    _inherit = 'res.partner'
+    associate_member_id = fields.Many2one('res.partner', 'Associate Member', help="A member with whom you want to associate your membership.It will consider the membership state of the associated member.", oldname='associate_member')
+    member_lines_ids = fields.One2many('membership.membership_line', 'partner_id', 'Membership', oldname='member_lines')
+    free_member = fields.Boolean(help="Select if you want to give free membership.")
+    membership_amount = fields.Float(digits=(16, 2), help='The price negotiated by the partner')
+    membership_state = fields.Selection(compute='_compute_get_membership_state', string='Current Membership Status', selection=membership.STATE, store=True,
```
Comment (atp-odoo) [inline addons/membership/models/partner.py]: function name should be : "_compute_membership_state"

File: addons/membership/views/membership_view.xml [hunk 13]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0"?>
+<?xml version="1.0" encoding="UTF-8"?>
 <openerp>
```
Comment (dja-odoo) [inline addons/membership/views/membership_view.xml]: We can use ``<odoo>`` tag and remove ``<openerp>`` and ``<data>``.

Comment (dja-odoo) [inline addons/membership/views/membership_view.xml]: Also remove oe_inline class from the file , and use o_row if needed.

**Comment by dja-odoo** [comment]
inside `tests/__init__.py`, use odoo in import statement instead of openerp.

File: addons/membership/__openerp__.py [hunk 14]
```diff
@@ -24,18 +24,16 @@
     'data': [
```
Comment (dja-odoo) [inline addons/membership/__openerp__.py]: You can remove version from this file.

File: addons/membership/models/__init__.py [hunk 15]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
```
Comment (dja-odoo) [inline addons/membership/models/__init__.py]: Few of the py files are having licence stuff while others aren't. Can you make it uniform ?

File: addons/membership/models/invoice.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+from datetime import date
+from odoo import api, models, fields, _
```
Comment (dja-odoo) [inline addons/membership/models/invoice.py]: Unused imports date, _, dp, ValidationError. Also order imports alphabetically (api, fields, models).

File: addons/membership/models/invoice.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+import odoo.addons.decimal_precision as dp
+from odoo.exceptions import ValidationError
+
+
+class Invoice(models.Model):
+    _inherit = 'account.invoice'
+
```
Comment (dja-odoo) [inline addons/membership/models/invoice.py]: Remove extra white-spaces.

File: addons/membership/models/invoice.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+        res = super(Invoice, self).action_cancel()
+        today_date = fields.Date.today()
+        membership_lines = self.env['membership.membership_line'].search([('account_invoice_line_id', 'in', self.mapped('invoice_line_ids').ids)])
+        membership_lines.write({'date_cancel': today_date})
+        return res
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
```
Comment (dja-odoo) [inline addons/membership/models/invoice.py]: Same as above.

File: addons/membership/models/invoice.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+        result = super(AccountInvoiceLine, self).create(vals)
+        membership_lines = MembershipLine.search([('account_invoice_line_id', '=', result.id)])
+        if result.invoice_id.type == 'out_invoice' and result.product_id.membership and not membership_lines:
+            # Product line is a membership product
+            date_from = result.product_id.membership_date_from
+            date_to = result.product_id.membership_date_to
+            if result.invoice_id.date_invoice > date_from and result.invoice_id.date_invoice < date_to:
+                date_from = result.invoice_id.date_invoice
+            MembershipLine.create({
+                        'partner_id': result.invoice_id.partner_id.id or False,
```
Comment (dja-odoo) [inline addons/membership/models/invoice.py]: no need to specify or condition.

File: addons/membership/models/membership.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    @api.depends('account_invoice_id.state', 'partner_id.membership_state')
+    def _compute_state(self):
+        """Compute the state lines """
+        for line in self:
+            if not line.account_invoice_id:
+                line.state = 'canceled'
+                continue
+            istate = line.account_invoice_id.state
+            state = 'none'
+            if (istate == 'draft') | (istate == 'proforma'):
```
Comment (dja-odoo) [inline addons/membership/models/membership.py]: Can't we simply write `` istate in ['draft', 'proforma'] `` ?

File: addons/membership/models/membership.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    ('paid', 'Paid Member'),
+]
+
+
+class MembershipLine(models.Model):
+    _name = 'membership.membership_line'
+    _description = __doc__
+    _rec_name = 'partner_id'
+    _order = 'id desc'
+    partner_id = fields.Many2one('res.partner', 'Partner', ondelete='cascade', index=1, oldname='partner')
```
Comment (dja-odoo) [inline addons/membership/models/membership.py]: Use string='Partner' convention wherever strings are provided for the fields(check all the model files).

File: addons/membership/models/membership.py [hunk 22]
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo import api, models, fields, _
+import odoo.addons.decimal_precision as dp
```
Comment (dja-odoo) [inline addons/membership/models/membership.py]: Order the imports according to [guidelines](https://www.odoo.com/documentation/9.0/reference/guidelines.html#imports).

File: addons/membership/models/product.py [hunk 23]
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+from odoo import api, fields, models, _
+import odoo.addons.decimal_precision as dp
+from odoo.exceptions import UserError
```
Comment (dja-odoo) [inline addons/membership/models/product.py]: Unused imports api, _, dp, UserError.

File: addons/membership/models/partner.py [hunk 24]
```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+from datetime import date
+import membership
+from odoo import api, models, fields, _
+import odoo.addons.decimal_precision as dp
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: Unused import dp.

File: addons/membership/models/membership.py [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+                    state = 'canceled'
+            elif istate == 'cancel':
+                state = 'canceled'
+            line.state = state
+
+    @api.constrains('date_to')
+    def _check_membership_date(self):
+        """Check if membership product is not in the past """
+
+        self.env.cr.execute('''
```
Comment (dja-odoo) [inline addons/membership/models/membership.py]: Probably this can be done without using query. Can you re-check ?

File: addons/membership/models/partner.py [hunk 26]
```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+from datetime import date
+import membership
+from odoo import api, models, fields, _
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: import order should be api, fields, model, _

File: addons/membership/models/partner.py [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+                                        '-Invoiced Member: A member whose invoice has been created.\n'
+                                        '-Paying member: A member who has paid the membership fee.')
+    membership_start = fields.Date(compute='_compute_membership_date', store=True, help='Date from which membership becomes active.')
+    membership_stop = fields.Date(compute='_compute_membership_date', store=True, help='Date until which membership remains active.')
+    membership_cancel = fields.Date(compute='_compute_membership_date', store=True, help='Date on which membership has been cancelled')
+
+    @api.depends('free_member', 'member_lines_ids')
+    def _compute_membership_date(self):
+        """Return  date of membership"""
+        if self.ids:
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: if self: isn't enough ?

File: addons/membership/models/partner.py [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+            level -= 1
+        return
+
+    @api.model
+    def _cron_update_membership(self):
+        partners = self.search([('membership_state', '=', 'paid')])
+        if partners:
+            partners._store_set_values(['membership_state'])
+
+    @api.multi
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: Do we need decorator ?

File: addons/membership/models/partner.py [hunk 29]
```diff
@@ -0,0 +1,173 @@
...
+    def _membership_state(self):
+        """This Function return Membership State For Given Partner. """
+        today_date = date.today()
+        for partner_data in self:
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: partner_data -> partner.

File: addons/membership/models/partner.py [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+            if partner_data.member_lines_ids:
+                for mline in partner_data.member_lines_ids:
+                    if fields.Date.from_string(mline.date_to) >= today_date:
+                        if mline.account_invoice_line_id and mline.account_invoice_line_id.invoice_id:
+                            mstate = mline.account_invoice_line_id.invoice_id.state
+                            if mstate == 'paid':
+                                state = 0
+                                inv = mline.account_invoice_line_id.invoice_id
+                                for payment in inv.payment_ids:
+                                    if payment.invoice_ids and any(inv.type == 'out_refund' for inv in payment.invoice_ids):
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: instead of above three lines, you can write 
```python
if mline.account_invoice_line_id.invoice_id.mapped('payment_ids.invoice_ids').filtered(lambda inv: inv.type == 'out_refund'):
```
And, check whether `mline.account_invoice_line_id.invoice_id` is same as `mline.account_invoice_id` or not. If they are same , you can use the latter one for this condition(only if you are pretty sure about it).

File: addons/membership/models/partner.py [hunk 31]
```diff
@@ -0,0 +1,173 @@
...
+                                break
+                            elif mstate == 'open' and state != 0:
+                                state = 1
+                            elif mstate == 'cancel' and state != 0 and state != 1:
+                                state = 2
+                            elif (mstate == 'draft' or mstate == 'proforma') and state != 0 and state != 1:
+                                state = 3
+                if state == 4:
+                    for mline in partner_data.member_lines_ids:
+                        if fields.Date.from_string(mline.date_from) < today_date and fields.Date.from_string(mline.date_to) < today_date and mline.date_from <= mline.date_to and (mline.account_invoice_line_id and mline.account_invoice_line_id.invoice_id.state) == 'paid':
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: ``mline.account_invoice_line_id.invoice_id.state == 'paid'`` is enough for last condition.

File: addons/membership/models/partner.py [hunk 32]
```diff
@@ -0,0 +1,173 @@
...
+                    partner_data.membership_state = 'old'
+                elif state == 6:
+                    partner_data.membership_state = 'none'
+            if partner_data.free_member and state != 0:
+                partner_data.membership_state = 'free'
+            if partner_data.associate_member_id:
+                partner_data.membership_state = partner_data.associate_member_id._membership_state()
+            return partner_data.membership_state
+
+    @api.multi
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: Do we need decorator here ?

File: addons/membership/models/partner.py [hunk 33]
```diff
@@ -0,0 +1,173 @@
...
+        """
+        AccountInv = self.env['account.invoice']
+        InvoiceLine = self.env['account.invoice.line']
+        product_id = product_id or datas.get('membership_product_id', False)
+        amount = datas.get('amount', 0.0)
+        invoice_list = []
+        for partner in self:
+            account_id = partner.property_account_receivable_id and partner.property_account_receivable_id.id or False
+            fpos_id = partner.property_account_position_id and partner.property_account_position_id.id or False
+            addr = partner.address_get(['invoice'])
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: ``account_id = partner.property_account_receivable_id.id`` and ``fpos_id = partner.property_account_position_id.id`` will do the job.

File: addons/membership/models/partner.py [hunk 34]
```diff
@@ -0,0 +1,173 @@
...
+            if partner.free_member:
+                raise UserError(_("Partner is a free Member."))
+            if not addr.get('invoice', False):
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: ``addr.get('invoice')`` isn't enough ?

File: addons/membership/models/partner.py [hunk 35]
```diff
@@ -0,0 +1,173 @@
...
+                raise UserError(_("Partner doesn't have an address to make the invoice."))
+
+            invoice_id = AccountInv.create({
+                'partner_id': partner.id,
+                'account_id': account_id,
+                'fiscal_position_id': fpos_id or False
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: No need for or condition.

File: addons/membership/models/partner.py [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+        """
+        level = 100
+        recursive_ids = self.ids
+        while len(recursive_ids):
+            self.env.cr.execute('SELECT DISTINCT associate_member_id FROM res_partner WHERE id IN %s', (tuple(recursive_ids),))
+            recursive_ids = filter(None, map(lambda x: x[0], self.env.cr.fetchall()))
+            if not level:
+                raise UserError(_('Error ! You cannot create recursive associated members.'))
+            level -= 1
+        return
```
Comment (dja-odoo) [inline addons/membership/models/partner.py]: Can we remove query ? 
```python
while len(recursive_ids):
    recursive_ids = self.search([('id', 'in', recursive_ids)]).mapped('associate_member_id').ids
    if not level:
        raise UserError(_('Error ! You cannot create recursive associated members.'))
    level -= 1
```

Comment (dja-odoo) [inline addons/membership/models/partner.py]: Probably the whole method can be improved like : 
```python
    @api.constrains('associate_member_id')
    def _check_associate_member(self):
        """Check  Recursive  for Associated Members.
        """
        for partner in self:
            if not partner._check_recursion(parent='associate_member_id'):
                raise UserError(_('Error ! You cannot create recursive associated members.'))
```
Can you check ?

File: addons/membership/report/report_membership.py [hunk 37]
```diff
@@ -1,43 +1,32 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields, osv
-from openerp import tools
-import openerp.addons.decimal_precision as dp
+from odoo.addons.membership.models import membership
+from odoo import api, fields, models, tools, _
+import odoo.addons.decimal_precision as dp
```
Comment (dja-odoo) [inline addons/membership/report/report_membership.py]: Unused imports api, _, dp

File: addons/membership/tests/test_membership.py [hunk 38]
```diff
@@ -60,7 +60,7 @@ def test_paid_membership(self):
         self.assertEqual(
```
Comment (dja-odoo) [inline addons/membership/tests/test_membership.py]: at line [#31](https://github.com/odoo/odoo/pull/8141/files#diff-c34058894ae4adbf78c3b9441bcfd107R31), remove index [0] . limit=1 is already there.

File: addons/membership/wizard/membership_invoice.py [hunk 39]
```diff
@@ -1,50 +1,49 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields, osv
-import openerp.addons.decimal_precision as dp
+from odoo import api, fields, models, _
+import odoo.addons.decimal_precision as dp
```
Comment (dja-odoo) [inline addons/membership/wizard/membership_invoice.py]: Remove unused imports.

File: addons/membership/wizard/membership_invoice.py [hunk 40]
```diff
@@ -1,50 +1,49 @@
...
 
-class membership_invoice(osv.osv_memory):
+
+class membership_invoice(models.TransientModel):
```
Comment (dja-odoo) [inline addons/membership/wizard/membership_invoice.py]: Class-name should be in CamelCase

File: addons/membership/wizard/membership_invoice.py [hunk 41]
```diff
@@ -1,50 +1,49 @@
...
     _name = "membership.invoice"
     _description = "Membership Invoice"
-    _columns = {
-        'product_id': fields.many2one('product.product','Membership', required=True),
-        'member_price': fields.float('Member Price', digits_compute= dp.get_precision('Product Price'), required=True),
-    }
-    def onchange_product(self, cr, uid, ids, product_id=False):
+    product_id = fields.Many2one('product.product', 'Membership', required=True)
+    member_price = fields.Float('Member Price', digits=dp.get_precision('Product Price'), required=True)
+
```
Comment (dja-odoo) [inline addons/membership/wizard/membership_invoice.py]: Extra white-spaces here.

File: addons/membership/wizard/membership_invoice.py [hunk 42]
```diff
@@ -1,50 +1,49 @@
...
-                'membership_product_id': data.product_id.id,
-                'amount': data.member_price
-            }
-        invoice_list = partner_obj.create_membership_invoice(cr, uid, context.get('active_ids', []), datas=datas, context=context)
+        pid = self.env.context.get('active_ids', [])
+        if self:
+            datas.update(membership_product_id=self.product_id.id, amount=self.member_price)
+        if pid:
+            partner_id = ResPartner.browse(pid)
+            invoice_list = partner_id.create_membership_invoice(datas=datas)
```
Comment (dja-odoo) [inline addons/membership/wizard/membership_invoice.py]: Use ``invoice_list = self.env['res.partner'].browse(pid).create_membership_invoice(datas=datas)``
No need for assignation of ResPartner and partner_id.

File: addons/membership/wizard/membership_invoice.py [hunk 43]
```diff
@@ -1,50 +1,49 @@
...
-            context = {}
-        data = self.browse(cr, uid, ids, context=context)
-        if data:
-            data = data[0]
-            datas = {
-                'membership_product_id': data.product_id.id,
-                'amount': data.member_price
-            }
-        invoice_list = partner_obj.create_membership_invoice(cr, uid, context.get('active_ids', []), datas=datas, context=context)
+        pid = self.env.context.get('active_ids', [])
```
Comment (dja-odoo) [inline addons/membership/wizard/membership_invoice.py]: self.env.context.get('active_ids') isn't enough ?

File: addons/membership/models/invoice.py [hunk 44]
```diff
@@ -0,0 +1,79 @@
...
+        result = super(AccountInvoiceLine, self).create(vals)
+        membership_lines = MembershipLine.search([('account_invoice_line', '=', result.id)])
+        if result.invoice_id.type == 'out_invoice' and result.product_id.membership and not membership_lines:
+            # Product line is a membership product
+            date_from = result.product_id.membership_date_from
+            date_to = result.product_id.membership_date_to
+            if result.invoice_id.date_invoice > date_from and result.invoice_id.date_invoice < date_to:
+                date_from = result.invoice_id.date_invoice
+            MembershipLine.create({
+                        'partner': result.invoice_id.partner_id.id or False,
```
Comment (atp-odoo) [inline addons/membership/models/invoice.py]: result.invoice_id.partner_id.id enough, no need of False.

File: addons/membership/models/membership.py (hunk 22 above)
Comment (atp-odoo) [inline addons/membership/models/membership.py]: :+1:

File: addons/membership/models/membership.py (hunk 20 above)
Comment (atp-odoo) [inline addons/membership/models/membership.py]: :+1:

File: addons/membership/models/partner.py [hunk 45]
```diff
@@ -0,0 +1,163 @@
...
+            partner.membership_state = 'none'
+            if partner.membership_cancel and today_date > fields.Date.from_string(partner.membership_cancel):
+                partner.membership_state = 'free' if partner.free_member else 'canceled'
+                continue
+            if partner.membership_stop and today_date > fields.Date.from_string(partner.membership_stop):
+                partner.membership_state = 'free' if partner.free_member else 'old'
+                continue
+
+            state = 4
+            if partner.member_lines:
```
Comment (atp-odoo) [inline addons/membership/models/partner.py]: no need to check this condition "if  partner.member_lines:"

File: addons/membership/models/partner.py [hunk 46]
```diff
@@ -0,0 +1,163 @@
...
+                for mline in partner.member_lines:
+                    if fields.Date.from_string(mline.date_to) >= today_date:
+                        if mline.account_invoice_line and mline.account_invoice_line.invoice_id:
```
Comment (atp-odoo) [inline addons/membership/models/partner.py]: if mline.account_invoice_line and mline.account_invoice_line.invoice_id: ==>  if mline.account_invoice_line.invoice_id:

File: addons/membership/report/report_membership_view.xml [hunk 47]
```diff
@@ -1,8 +1,7 @@
-<?xml version="1.0"?>
-<openerp>
-    <data>
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
         <!-- REPORTING/MEMBERSHIP BY YEAR  -->
-        <record model="ir.ui.view" id="view_report_membership_search">
+        <record id="view_report_membership_search" model="ir.ui.view">
```
Comment (atp-odoo) [inline addons/membership/report/report_membership_view.xml]: revert this change not required...

File: addons/membership/models/membership.py [hunk 48]
```diff
@@ -20,50 +15,43 @@
 ]
 
 
-class membership_line(osv.osv):
+class MembershipLine(models.Model):
     _name = 'membership.membership_line'
     _description = __doc__
+    _rec_name = 'partner'
+    _order = 'id desc'
```
Comment (tivisse) [inline addons/membership/models/membership.py]: Why changing the rec_name and the _order?

Comment (tivisse) [inline addons/membership/models/membership.py]: Forget about it.

File: addons/membership/models/membership.py [hunk 49]
```diff
@@ -20,50 +15,43 @@
...
+             "-Old Member: A member whose membership date has expired.\n"
+             "-Waiting Member: A member who has applied for the membership and whose invoice is going to be created.\n"
+             "-Invoiced Member: A member whose invoice has been created.\n"
+             "-Paid Member: A member who has paid the membership amount.")
 
-    def _state(self, cr, uid, ids, name, args, context=None):
+    @api.depends('account_invoice_line', 'account_invoice_line.invoice_id',
+                 'account_invoice_line.invoice_id.state',
+                 'account_invoice_line.invoice_id.payment_ids',
+                 'account_invoice_line.invoice_id.payment_ids.invoice_ids')
```
Comment (tivisse) [inline addons/membership/models/membership.py]: I think that it's false because: 
- You don't use the second trigger method and model
- For the first trigger condition, you use too many fields. 

I think that this is sufficient. 
``` python
    @api.depends('account_invoice_line.invoice_id.state','partner.membership_state')
```

File: addons/membership/models/partner.py [hunk 50]
```diff
@@ -1,253 +1,188 @@
...
+        help="Date on which membership has been cancelled")
+
+    @api.depends('membership_stop', 'associate_member',
+                 'free_member', 'member_lines',
+                 'member_lines.date_to', 'member_lines.date_from',
+                 'member_lines.account_invoice_line',
+                 'member_lines.account_invoice_line.invoice_id',
+                 'member_lines.account_invoice_line.invoice_id.state',
+                 'member_lines.account_invoice_line.invoice_id.payment_ids')
+    def _compute_membership_state(self):
```
Comment (tivisse) [inline addons/membership/models/partner.py]: Why did you put so many fields in the depends. Once again, I think that 
``` python
    @api.depends('member_lines.account_invoice_line.invoice_id.state',
                 'member_lines.account_invoice_line.invoice_id.invoice_line_ids',
                 'member_lines.account_invoice_line.invoice_id.payment_ids',
                 'member_lines.state',
                 'free_member',
                 'membership_state',
                 'associate_member')
```
is sufficient, do you agree ?

File: addons/membership/models/partner.py [hunk 51]
```diff
@@ -1,253 +1,188 @@
...
-            if partner_data.membership_stop and today > partner_data.membership_stop:
-                res[id] = 'free' if partner_data.free_member else 'old'
+            if partner.membership_stop and today > partner.membership_stop:
+                res[partner.id] = 'free' if partner.free_member else 'old'
                 continue
 
             s = 4
-            if partner_data.member_lines:
-                for mline in partner_data.member_lines:
+            res[partner.id] = 'none'
```
Comment (tivisse) [inline addons/membership/models/partner.py]: Why putting this statement here, you erase what you've done before in the loop

File: addons/membership/report/report_membership.py [hunk 52]
```diff
@@ -15,35 +13,35 @@
...
+    tot_earned = fields.Float(string='Earned Amount', digits=0, readonly=True)
+    partner_id = fields.Many2one('res.partner', string='Member', readonly=True)
+    associate_member_id = fields.Many2one('res.partner', string='Associate Member', readonly=True)
+    membership_id = fields.Many2one('product.product', string='Membership Product', readonly=True)
+    membership_state = fields.Selection(STATE, string='Current Membership State', readonly=True)
+    user_id = fields.Many2one('res.users', string='Salesperson', readonly=True)
+    company_id = fields.Many2one('res.company', string='Company', readonly=True)
+    quantity = fields.Integer(readonly=True)
+
+    def init(self):
```
Comment (tivisse) [inline addons/membership/report/report_membership.py]: @api.model_cr

File: addons/membership/models/partner.py (hunk 50 above)
Comment (rgo-odoo) [inline addons/membership/models/partner.py]: when there is chain like `member_lines.account_invoice_line.invoice_id.payment_ids`(or any such chain is used inside related compute field), I thought it is required to trigger store when there is a change in any of the fields of the chain. That is why I use to keep all the fields(starting from `member_lines`) separately in depends. So, I am not quite really sure if it is sufficient. May be we can have confirmation from @rco-odoo as these functions has bit complex dependencies.
But IMO, we should keep `membership_stop` in the above list as it is being used inside `_membership_state()`

Comment (tivisse) [inline addons/membership/models/partner.py]: Yep, it's a little tricky and this module is quite badly written. I'll check with him.

File: addons/membership/models/partner.py (hunk 51 above)
Comment (rgo-odoo) [inline addons/membership/models/partner.py]: nope, actually there is `continue` statement with each `if` above
